README.md
![Chums Logo](https://intranet.chums.com/images/chums/chums-badge-120x120.png)
# GTIN Tools

## Installation
    npm install @chumsinc/gtin-tools

## Usage & Examples

```typescript
import {
    couldBeGTIN, 
    trimUPC, 
    parseCheckDigit, 
    parseUPCA, 
    parseGTINFormat,
    formatGTIN
} from "@chumsinc/gtin-tools";

couldBeGTIN('ASD123') //returns false
couldBeGTIN('01234567890') //returns true
couldBeGTIN('0 12345 67890') //returns true
couldBeGTIN('0 12345 67890 5') //returns true
couldBeGTIN('0 12345 67890 X') //returns false

trimUPC('0 12345 67890 5') // returns '012345678905'
parseCheckDigit('01234567890', 'GTIN-12') // returns '5'

parseGTINFormat('01234567890') // returns 'GTIN-12', gtin-12 without check digit
parseGTINFormat('012345678905') // returns 'GTIN-12'
parseGTINFormat('451234567890') // returns 'GTIN-13' Japan GTIN-13 without check digit

formatGTIN('01234567890') // returns '0 12345 67890 5', corrects check digit & adds spaces 
formatGTIN('012345678908') // returns '0 12345 67890 5', corrects check digit & adds spaces 
formatGTIN('012345678908', true) // returns '012345678905', corrects check digit 

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v1.0.2](https://github.com/UtahGooner/gtin-tools/compare/v1.0.1...v1.0.2)

### Commits

- feat: Add README.md [`4d7ca79`](https://github.com/UtahGooner/gtin-tools/commit/4d7ca794913df727a1dacce8dfdb84cc566fd9bd)
- fix: adding exports to package.json [`60ff7e6`](https://github.com/UtahGooner/gtin-tools/commit/60ff7e638952528e73a0479ed5cbe0dd98c7ca35)
- feat: Added version script. [`3c2bc31`](https://github.com/UtahGooner/gtin-tools/commit/3c2bc317136331a5f5411ed748e273ffc44dda67)

## v1.0.1 - 2023-01-31

### Commits

- feat: Added a changelog. [`01b0a3a`](https://github.com/UtahGooner/gtin-tools/commit/01b0a3a82071ac9d34c868f0dd158e3d4f6a6604)
- Initial commit [`ac939cf`](https://github.com/UtahGooner/gtin-tools/commit/ac939cf639e7517c25f9d52fa134053a59e929ee)
- Initial Commit [`fe21290`](https://github.com/UtahGooner/gtin-tools/commit/fe212908e3bc6c50808e5277d960eb0d69d8caf0)
- feat: Added version script. [`11b9c24`](https://github.com/UtahGooner/gtin-tools/commit/11b9c240920821d707e1a5ba9a960984358c9f02)


main.js
export const barcodeDigits = {
    // ['GTIN-8']: 8,
    ['GTIN-12']: 12,
    ['GTIN-13']: 13,
    ['GTIN-14']: 14,
    ['GSIN']: 17,
    ['SSCC']: 18,
};
export const barcodeSplits = {
    ['GTIN-12']: /(\d)(\d{5})(\d{5})(\d)/,
    ['GTIN-13']: /(\d)(\d{6})(\d{6})/,
    ['GTIN-14']: /(\d)(\d{2})(\d{5})(\d{5})(\d)/,
};
export const couldBeGTIN = (gtin) => /^[0-9 ]+$/.test(gtin);
export function trimUPC(digits) {
    return digits.trim().replace(/\D/g, '');
}
export function parseCheckDigit(str, format = 'GTIN-12') {
    const digitSum = [0, 0];
    let digits = trimUPC(str);
    const stdDigits = barcodeDigits[format];
    if (!digits || (digits.length < (stdDigits - 1)) || (digits.length > stdDigits)) {
        return '';
    }
    else if (digits.length === stdDigits || digits.length === stdDigits - 1) {
        digits = digits.slice(0, stdDigits - 1);
    }
    else {
        return '';
    }
    const checkSum = digits.split('')
        .reverse()
        .map((digit, index) => index % 2 === 0 ? Number(digit) * 3 : Number(digit))
        .reduce((pv, cv) => pv + cv, 0);
    return ((10 - checkSum % 10) % 10).toString();
}
export function parseUPCA(upc) {
    const digits = trimUPC(upc);
    if (!digits || digits.length < 11 || digits.length > 12) {
        return upc;
    }
    return digits.slice(0, 11) + parseCheckDigit(digits, 'GTIN-12');
}
export function parseGTINFormat(str) {
    const digits = trimUPC(str);
    switch (digits.length) {
        case barcodeDigits["GTIN-12"] - 1:
            return 'GTIN-12';
        case barcodeDigits['GTIN-12']:
            if (/^4[59]/.test(digits)) {
                return 'GTIN-13';
            }
            return 'GTIN-12';
        case barcodeDigits['GTIN-13']:
            return 'GTIN-13';
        case barcodeDigits['GTIN-14']:
            return 'GTIN-14';
        case barcodeDigits.GSIN:
            return 'GSIN';
        case barcodeDigits.SSCC:
            return 'SSCC';
        default:
            return null;
    }
}
export function formatUPC(upc, raw = false) {
    if (!couldBeGTIN(upc)) {
        return upc;
    }
    const format = parseGTINFormat(upc);
    if (!format) {
        return upc;
    }
    const digits = trimUPC(upc).slice(0, barcodeDigits[format] - 1) + parseCheckDigit(upc, format);
    if (raw) {
        return digits;
    }
    const splitRegExp = barcodeSplits[format];
    if (!splitRegExp) {
        return digits;
    }
    switch (format) {
        case 'GTIN-12':
            const gtin12 = splitRegExp.exec(digits);
            if (!gtin12) {
                return upc;
            }
            return `${gtin12[1]} ${gtin12[2]} ${gtin12[3]} ${gtin12[4]}`;
        case 'GTIN-13':
            const gtin13 = splitRegExp.exec(digits);
            if (!gtin13) {
                return upc;
            }
            return `${gtin13[1]} ${gtin13[2]} ${gtin13[3]}`;
        case 'GTIN-14':
            const gtin14 = splitRegExp.exec(digits);
            if (!gtin14) {
                return upc;
            }
            return `${gtin14[1]} ${gtin14[2]} ${gtin14[3]} ${gtin14[4]} ${gtin14[5]}`;
        default:
            return digits;
    }
}
//# sourceMappingURL=main.js.map

main.ts
export type BarcodeType = 'GTIN-12' | 'GTIN-13' | 'GTIN-14' | 'GSIN' | 'SSCC';

export type BarcodeTypeDigits = {
    [key in BarcodeType]: number;
};

export type BarcodeTypeSplits = {
    [key in BarcodeType]?: RegExp;
}


export const barcodeDigits:BarcodeTypeDigits = {
    // ['GTIN-8']: 8,
    ['GTIN-12']: 12,
    ['GTIN-13']: 13,
    ['GTIN-14']: 14,
    ['GSIN']: 17,
    ['SSCC']: 18,
}

export const barcodeSplits:BarcodeTypeSplits = {
    ['GTIN-12']: /(\d)(\d{5})(\d{5})(\d)/,
    ['GTIN-13']: /(\d)(\d{6})(\d{6})/,
    ['GTIN-14']: /(\d)(\d{2})(\d{5})(\d{5})(\d)/,
}

export const couldBeGTIN = (gtin:string) => /^[0-9 ]+$/.test(gtin);

export function trimUPC(digits:string):string {
    return digits.trim().replace(/\D/g, '');
}


export function parseCheckDigit(str: string, format: BarcodeType = 'GTIN-12'):string {
    let digits = trimUPC(str);
    const stdDigits = barcodeDigits[format];
    if (!digits || (digits.length < (stdDigits - 1)) || (digits.length > stdDigits)) {
        return '';
    } else if (digits.length === stdDigits || digits.length === stdDigits - 1) {
        digits = digits.slice(0, stdDigits - 1);
    } else {
        return '';
    }

    const checkSum = digits.split('')
        .reverse()
        .map((digit, index) => index % 2 === 0 ? Number(digit) * 3 : Number(digit))
        .reduce((pv, cv) => pv + cv, 0);

    return ((10 - checkSum % 10) % 10).toString();
}


export function parseUPCA(upc:string):string {
    const digits = trimUPC(upc);
    if (!digits || digits.length < 11 || digits.length > 12) {
        return upc;
    }
    return digits.slice(0, 11) + parseCheckDigit(digits, 'GTIN-12');
}

export function parseGTINFormat(str:string):BarcodeType|null {
    const digits = trimUPC(str);
    switch (digits.length) {
    case barcodeDigits["GTIN-12"] - 1:
        return 'GTIN-12';
    case barcodeDigits['GTIN-12']:
        if (/^4[59]/.test(digits)) {
            return 'GTIN-13';
        }
        return 'GTIN-12';
    case barcodeDigits['GTIN-13']:
        return 'GTIN-13';
    case barcodeDigits['GTIN-14']:
        return 'GTIN-14';
    case barcodeDigits.GSIN:
        return 'GSIN';
    case barcodeDigits.SSCC:
        return 'SSCC';
    default:
        return null;
    }
}

export function formatGTIN(upc:string, raw:boolean = false) {
    if (!couldBeGTIN(upc)) {
        return upc;
    }

    const format = parseGTINFormat(upc);
    if (!format) {
        return upc;
    }
    const digits = trimUPC(upc).slice(0, barcodeDigits[format] - 1) + parseCheckDigit(upc, format);
    if (raw) {
        return digits;
    }

    const splitRegExp = barcodeSplits[format];
    if (!splitRegExp) {
        return digits;
    }

    switch (format) {
    case 'GTIN-12':
        const gtin12 = splitRegExp.exec(digits);
        if (!gtin12) {
            return upc;
        }
        return `${gtin12[1]} ${gtin12[2]} ${gtin12[3]} ${gtin12[4]}`;
    case 'GTIN-13':
        const gtin13 = splitRegExp.exec(digits);
        if (!gtin13) {
            return upc;
        }
        return `${gtin13[1]} ${gtin13[2]} ${gtin13[3]}`;

    case 'GTIN-14':
        const gtin14 = splitRegExp.exec(digits);
        if (!gtin14) {
            return upc;
        }
        return `${gtin14[1]} ${gtin14[2]} ${gtin14[3]} ${gtin14[4]} ${gtin14[5]}`;

    default:
        return digits;
    }
}
